Swift concurrency

rcasula opened the pull request:
Update the Library to the new swift-concurrency standard.
I was thinking if removing the composable architecture dependency makes sense.

This PR is still in draft because I have to:
- [ ] update documentation
- [x] eventually remove the composable-architecture dependency

Please let me know if this PR is any good!

rcasula committed:
Convert functions to async/await

rcasula committed:
Make use of XCTUnimplemented

rcasula committed:
Convert LocationManagerDelegate to use continuation instead of Effect

rcasula committed:
Convert to async/await

rcasula committed:
Remove old deprecations

stephencelis commented:
@rcasula Thanks for looking into this! We definitely have plans to adopt this new style soon, but haven't had time to look into it. This looks good! Just a couple things we need to figure out in the future:

1. Does the `manager` need to be protected in an actor? And what concurrency warnings show up when we turn on `-Xfrontend -warn-concurrency -Xfrontend -enable-actor-data-race-checks`?
2. We'd probably want to coordinate this change with a library rename. Something like `core-location-client` or `core-location-dependency`.

Going to keep this open for now while we find the time to figure out those details. Thanks for getting this work started though!

rcasula committed:
Fix old code

rcasula committed:
Add dependencies

rcasula committed:
Update composable architecture

rcasula committed:
Remove composable-architecture dependency

3a4oT commented:
Hey, what is the status of this PR? What needs to be done to wrap up the new release?

johankool committed:
Update dependencies

rcasula committed:
Update dependencies

p4checo committed:
Run `live` on the MainActor, allow multiple `delegate` streams

When testing the current (Swift Concurrency) approach, the delegate
method calls were not being invoked, and after some investigation it
seems the problem was related with the `RunLoop` in which the
`CLLocationManager` was created. With the latest `Dependencies`/`TCA`
implementations, dependencies' access isn't bound to the `@MainActor`,
so code is run from arbitrary queues/threads spanning multiple RunLoops.

According to `CLLocationManager`'s
[documentation](https://developer.apple.com/documentation/corelocation/cllocationmanager):

> "Core Location calls the methods of your delegate object using the
> `RunLoop` of the thread on which you initialized the
> `CLLocationManager` object.

By forcing the `LocationManager.live`'s `CLLocationManager` instance to
be instantiated on the main `RunLoop` via `@MainActor`, the delegate
methods are correctly invoked. To minimize problems, all `.live`
instance APIs' closures were annotated with `@MainActor`.

Furthermore, as the `Dependencies` instances are shared and can be
called from multiple places concurrently, the `LocationManagerDelegate`
was updated to allow forwarding delegate calls to multiple streams
provided via `delegate()`.

Ran `swift-format` and performed some cleanups.

p4checo committed:
Enable and fix concurrency warnings, add `Sendable` conformances

- Enable swift concurrency warnings in `Package.swift`
- Add `Sendable` conformance where needed, and wrap `CLLocation` types
in `@UncheckedSendable` until properly annotated.

p4checo commented on Sources/ComposableCoreLocation/Live.swift:
```diff
@@ -47,14 +35,20 @@ extension LocationManager {
-            manager.dismissHeadingCalibrationDisplay()
-          #endif
+      delegate: {
+        AsyncStream { continuation in
```
In the meantime, opened MR rcasula/composable-core-location#2 which hopefully (also) addresses this issue 🤓

p4checo committed:
Comment compiler concurrency warnings from Package.swift

rcasula committed:
Restore code after rebase

rcasula committed:
Revert @MainActor call cause the method cannot be used on the main thread

rcasula committed:
Migrate example to TCA 1.0

p4checo commented:
For completeness' sake, following a discussion in this [PF's slack thread](https://pointfreecommunity.slack.com/archives/C04L2D0MNJH/p1693911737867579), to ensure `CLLocationManager` is accessed from the main `RunLoop` an alternative solution could be to use a `MainActorIsolated` structure as introduced [here](https://github.com/pointfreeco/swift-dependencies/discussions/22#discussioncomment-4681533). 

It appears to ensure `CLLocationManager`'s APIs are invoked on the Main RunLoop, so it has the benefit of avoiding the `Task { @MainActor }` and not requiring sprinkling `@MainActor` in all API closures, making it less verbose and possibly less error prone as well if more APIs are added in the future.

Seems like an interesting approach, but ideally someone with more experience in swift concurrency can chime in, or at least we have one more approach to consider 🤓

p4checo commented:
@rcasula can this MR be marked as ready for review? What documentation updates are missing? Thanks!

3a4oT commented on Package.swift:
```diff
@@ -34,3 +34,13 @@ let package = Package(
+//      "-Xfrontend", "-enable-actor-data-race-checks",
+//    ])
+//  )
+//}
```
Should it be enabled?

p4checo replied:
If we do, SPM complains about unsafe flags being used when integrating it in other packages/projects

3a4oT replied:
ohh, that's a good point. Just thinking loud here. What if we introduce ENV varible for local development and CI.
Something like :

```swift
if ProcessInfo.processInfo.environment["COMPOSABLE_DEVELOPER"] == "1" {
settings.append(.unsafeFlags([
     "-Xfrontend",
     "-warn-concurrency",
     "-Xfrontend",
     "-enable-actor-data-race-checks"
     ]))
}
```

Little incovinience to define variable locally but it can be documented on the README for developers.

p4checo replied:
That's definitely an interesting idea! Let's see what the rest have to say 🤓

rcasula committed:
Fix compilation errors with watchOS and macOS

yosuke1985 commented:
@stephencelis
I want my project upgraded to over TCA 1.0, but I cannot because of this project,
can you merge this pull request?

3a4oT commented:
Hello, any chances to proceed with it?

rcasula committed:
Rename library to CoreLocationClient

spiridonkopicl commented:
Thank you @rcasula, you saved my day 🍻
